DESIGN.md

```diff
@@ -0,0 +1,471 @@
+```bash
++------------------+
+|  Jobcontroller   |
+|   `joblib`       |
+|  (writes to disk)|
++------------------+
+         |
+         v
++------------------+
+|   stdout/stderr  |
+|   written to     |
+|   disk file      |
++------------------+
+         |
+         v
++------------------------------+
+|   On client StreamOutput()   |
+|   request:                   |
+|                              |
+|   - start goroutine:         |
+|     read from file           |
+|     and send to client       |
++------------------------------+
+         |
+         v
++------------------+
+|  gRPC Stream     |
+|   Response       |
++------------------+
+         |
+         v
++------------------+
+|     Client       |
++------------------+
+```
+##### `StreamOutput`
+- **Initial note**: The majority of this logic will be implemented in the `joblib` but `StreamOutput` will have this functionality.
+- `StreamOutput` will **only** stream from job output that is written to disk. Stdout and Stderr will be written to disk by JobController via `joblib` itself upon invocation of `StartJob`. A closed file descriptor (stdout/stderr) will signify the end of the file/stream.
+- `StreamOutput` will, after reading to EOF, block on inotify (Linux) and resume reading; no busy polling
+- Each `StreamOutput` call runs in its own goroutine. The handler reads the job’s persisted log file and writes chunks directly to the gRPC stream. We rely on gRPC flow control for backpressure and `context.Context` for cancellation. No intermediate per-client channels are used. Due to relying on `context.Context`, we can assure that no unneeded reading occurs (**joblib**)
+-  `Replay`: A client will be able to stream any output of a job even after the job has exited. (**joblib**)
+
+##### StreamOutput Requirements
+- Clients stream from the **beginning of the output**
+- Support **multiple clients** connecting at different times.
+- Avoid race conditions when jobs are writing and clients are reading.
+- Designed for **CLI usage** via a single `StreamOutput` gRPC call.
+
+
+### 2. `joblib` (Core Library)
+
+| Function             | Description                                             |
+| -------------------- | ------------------------------------------------------- |
+| `Start()`            | Starts a process via setpgid(), takes in an executable and args. Sets job status, cgroup limits and executes the process. See **Start** for details.  It also sets up the process' stdout and stderr and to be read from and written to files on disk. These file names should map up to the job id assigned and also on disk as well. Then wait for job to finish, or be terminated, get its exit status and set it job status as well.
+| `Stop()`             | Sends `SIGTERM` to the process group for as an initial try. If the process (or children) is still running after the timeout, it writes `1` to `/sys/fs/cgroup/jobs/<jobid>/cgroup.kill` to forcefully terminate it and remove the cgroup.
```

Comment:
Feel free to skip straight to `cgroup.kill` in order to simplify the implementation.